### PR TITLE
Adds developer documentation to readthedocs

### LINKED ISSUE #135: Add design documentation for overall code structure
Per @white238 's suggestion, I should add a page to the sphinx docs outlining how all of the classes are related.

### PR DESCRIPTION
**Pull request by jamiebramwell**
This adds a first draft of "user" and "developer" documentation to the readthedocs page. A copy of the documentation generated by this PR is available [here](https://serac.readthedocs.io/en/docs-bramwell-user_docs/sphinx/developer_guide.html).

Partially resolves https://github.com/LLNL/serac/issues/135.

### CODE REVIEW

File: src/docs/sphinx/developer_guide.rst [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+#. Override the virtual ``advanceTimestep()`` method. This should solve the discretized PDE based on the chosen time integration method. This often requires defining ``mfem::Operators`` to use MFEM-based nonlinear and time integration methods. 
+
+Important Data Structures
+-------------------------
+
+* `BasePhysics <../doxygen/html/classserac_1_1BasePhysics.html>`_: Interface class for a generic PDE simulation module.
+* `BoundaryCondition <../doxygen/html/classserac_1_1BoundaryCondition.html>`_: Class for storage of boundary condition-related data.
+* `BoundaryConditionManager <../doxygen/html/classserac_1_1BoundaryConditionManager.html>`_: Storage class for related boundary conditions.
+* `EquationSolver <../doxygen/html/classserac_1_1EquationSolver.html>`_: Class for solving nonlinear and linear discretized systems of equations.
+* `FiniteElementState <../doxygen/html/classserac_1_1FiniteElementState.html>`_: Data structure for storing a state variable discretized by MFEM and therefore the finite element method.
```
Comment (samuelpmishLLNL) [inline src/docs/sphinx/developer_guide.rst]: the wording on this line is unusual, maybe something like

```
Data structure describing a solution field and its underlying finite element discretization.
```

File: src/docs/sphinx/developer_guide.rst [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+#. Override the virtual ``completeSetup()`` method. This should include construction of all of the data structures needed for advancing the timestep of the PDE.
+#. Override the virtual ``advanceTimestep()`` method. This should solve the discretized PDE based on the chosen time integration method. This often requires defining ``mfem::Operators`` to use MFEM-based nonlinear and time integration methods. 
+
+Important Data Structures
+-------------------------
+
+* `BasePhysics <../doxygen/html/classserac_1_1BasePhysics.html>`_: Interface class for a generic PDE simulation module.
+* `BoundaryCondition <../doxygen/html/classserac_1_1BoundaryCondition.html>`_: Class for storage of boundary condition-related data.
+* `BoundaryConditionManager <../doxygen/html/classserac_1_1BoundaryConditionManager.html>`_: Storage class for related boundary conditions.
+* `EquationSolver <../doxygen/html/classserac_1_1EquationSolver.html>`_: Class for solving nonlinear and linear discretized systems of equations.
```
Comment (samuelpmishLLNL) [inline src/docs/sphinx/developer_guide.rst]: I think system of equations already implies discrete, 
```
Class for solving linear and nonlinear systems of equations.
```

**Review by samuelpmishLLNL — APPROVED** [review]
Looks good! I tried all the links, and they seem to go to the correct places.

**Review by joshessman-llnl — APPROVED** [review]
This is great!  I think we just need to remember to update this whenever significant changes are introduced.

**Comment by jamiebramwell** [comment]
LGTM

**Pull request merged**